// react-compare-image/dist/ReactCompareImage.mjs
import { useCallback as e, useEffect as t, useLayoutEffect as n, useRef as r, useState as i } from "react";
import { Fragment as a, jsx as o, jsxs as s } from "react/jsx-runtime";
//#region node_modules/tiny-invariant/dist/esm/tiny-invariant.js
var c = process.env.NODE_ENV === "production", l = "Invariant failed";
function u(e, t) {
	if (!e) {
		if (c) throw Error(l);
		var n = typeof t == "function" ? t() : t, r = n ? `${l}: ${n}` : l;
		throw Error(r);
	}
}
//#endregion
//#region src/hooks/useContainerWidth.ts
function d(e) {
	let t = r(null);
	return n(() => {
		if (!t.current) return;
		let n = new ResizeObserver((t) => {
			u(t[0], "ResizeObserver should have at least one entry"), e(t[0].contentRect.width);
		});
		return n.observe(t.current), e(t.current.getBoundingClientRect().width), () => {
			n.disconnect();
		};
	}, [e]), t;
}
//#endregion
//#region src/utils/calculateContainerHeight.ts
function f(e, t, n, r) {
	return e * (r === "taller" ? Math.max(t, n) : Math.min(t, n));
}
//#endregion
//#region src/utils/getImageRatio.ts
function p(e) {
	return e.naturalHeight / e.naturalWidth;
}
//#endregion
//#region src/ReactCompareImage.tsx
var m = (n) => {
	let { aspectRatio: c = "taller", handle: l = null, handleSize: u = 40, hover: m = !1, leftImage: h, leftImageAlt: g = "", leftImageCss: _ = {}, leftImageLabel: v = null, onSliderPositionChange: y, rightImage: b, rightImageAlt: x = "", rightImageCss: S = {}, rightImageLabel: C = null, skeleton: w = null, sliderLineColor: T = "#ffffff", sliderLineWidth: E = 2, sliderPositionPercentage: D = .5, vertical: O = !1 } = n, k = !O, [A, j] = i(D), [M, N] = i(!1), [P, F] = i(0), [I, L] = i(0), R = d((e) => F(e)), z = r(null), B = r(null), [V, H] = i(!1), U = e(() => {
		B.current?.complete && z.current?.complete && H(!0);
	}, []);
	t(() => (U(), () => {
		H(!1);
	}), [
		h,
		b,
		U
	]), t(() => {
		!(B.current && z.current) || P === 0 || !V || L(f(P, p(B.current), p(z.current), c));
	}, [
		P,
		V,
		c
	]), t(() => {
		if (!R.current || !V) return;
		let e = (e) => {
			if (!R.current) return;
			let t = R.current.getBoundingClientRect(), n = k ? e.clientX - t.left : e.clientY - t.top, r = E / 2, i = k ? P - r : I - r, a = Math.min(Math.max(n, r), i) / (k ? P : I);
			j(a), y?.(a);
		}, t = (t) => {
			N(!0), t.preventDefault(), i.setPointerCapture(t.pointerId), e(t);
		}, n = (t) => {
			(i.hasPointerCapture(t.pointerId) || m) && e(t);
		}, r = () => {
			N(!1);
		}, i = R.current;
		return i.addEventListener("pointerdown", t), i.addEventListener("pointermove", n), i.addEventListener("pointerup", r), () => {
			i.removeEventListener("pointerdown", t), i.removeEventListener("pointermove", n), i.removeEventListener("pointerup", r);
		};
	}, [
		V,
		c,
		I,
		P,
		k,
		m,
		E,
		R
	]);
	let W = {
		container: {
			boxSizing: "border-box",
			position: "relative",
			width: "100%",
			height: `${I}px`,
			overflow: "hidden",
			touchAction: k ? "pan-y" : "pan-x"
		},
		rightImage: {
			clipPath: k ? `inset(0px 0px 0px ${P * A}px)` : `inset(${I * A}px 0px 0px 0px)`,
			display: "block",
			height: "100%",
			objectFit: "cover",
			position: "absolute",
			width: "100%",
			...S
		},
		leftImage: {
			clipPath: k ? `inset(0px ${P * (1 - A)}px 0px 0px)` : `inset(0px 0px ${I * (1 - A)}px 0px)`,
			display: "block",
			height: "100%",
			objectFit: "cover",
			position: "absolute",
			width: "100%",
			..._
		},
		slider: {
			alignItems: "center",
			display: "flex",
			justifyContent: "center",
			position: "absolute",
			...k ? {
				cursor: "ew-resize",
				flexDirection: "column",
				height: "100%",
				left: `${P * A - u / 2}px`,
				top: 0,
				width: `${u}px`
			} : {
				cursor: "ns-resize",
				flexDirection: "row",
				height: `${u}px`,
				left: 0,
				top: `${I * A - u / 2}px`,
				width: "100%"
			}
		},
		line: {
			background: T,
			boxShadow: "0px 3px 1px -2px rgba(0, 0, 0, 0.2), 0px 2px 2px 0px rgba(0, 0, 0, 0.14), 0px 1px 5px 0px rgba(0, 0, 0, 0.12)",
			flex: "0 1 auto",
			height: k ? "100%" : `${E}px`,
			width: k ? `${E}px` : "100%"
		},
		handleCustom: {
			alignItems: "center",
			boxSizing: "border-box",
			display: "flex",
			flex: "1 0 auto",
			height: "auto",
			justifyContent: "center",
			width: "auto"
		},
		handleDefault: {
			alignItems: "center",
			border: `${E}px solid ${T}`,
			borderRadius: "100%",
			boxShadow: "0px 3px 1px -2px rgba(0, 0, 0, 0.2), 0px 2px 2px 0px rgba(0, 0, 0, 0.14), 0px 1px 5px 0px rgba(0, 0, 0, 0.12)",
			boxSizing: "border-box",
			display: "flex",
			flex: "1 0 auto",
			height: `${u}px`,
			justifyContent: "center",
			width: `${u}px`,
			transform: k ? "none" : "rotate(90deg)"
		},
		leftArrow: {
			border: `inset ${u * .15}px rgba(0,0,0,0)`,
			borderRight: `${u * .15}px solid ${T}`,
			height: "0px",
			marginLeft: `-${u * .25}px`,
			marginRight: `${u * .25}px`,
			width: "0px"
		},
		rightArrow: {
			border: `inset ${u * .15}px rgba(0,0,0,0)`,
			borderLeft: `${u * .15}px solid ${T}`,
			height: "0px",
			marginRight: `-${u * .25}px`,
			width: "0px"
		},
		leftLabel: {
			background: "rgba(0, 0, 0, 0.5)",
			color: "white",
			left: k ? "5%" : "50%",
			opacity: +!M,
			padding: "10px 20px",
			position: "absolute",
			top: k ? "50%" : "3%",
			transform: k ? "translate(0,-50%)" : "translate(-50%, 0)",
			transition: "opacity 0.1s ease-out"
		},
		rightLabel: {
			background: "rgba(0, 0, 0, 0.5)",
			color: "white",
			opacity: +!M,
			padding: "10px 20px",
			position: "absolute",
			...k ? {
				right: "5%",
				top: "50%",
				transform: "translate(0,-50%)"
			} : {
				left: "50%",
				bottom: "3%",
				transform: "translate(-50%, 0)"
			},
			transition: "opacity 0.1s ease-out"
		},
		leftLabelContainer: {
			clipPath: k ? `inset(0px ${P * (1 - A)}px 0px 0px)` : `inset(0px 0px ${I * (1 - A)}px 0px)`,
			height: "100%",
			position: "absolute",
			width: "100%"
		},
		rightLabelContainer: {
			clipPath: k ? `inset(0px 0px 0px ${P * A}px)` : `inset(${I * A}px 0px 0px 0px)`,
			height: "100%",
			position: "absolute",
			width: "100%"
		}
	};
	return /* @__PURE__ */ s(a, { children: [w && !V && /* @__PURE__ */ o("div", {
		style: { ...W.container },
		children: w
	}), /* @__PURE__ */ s("div", {
		style: {
			...W.container,
			display: V ? "block" : "none"
		},
		ref: R,
		"data-testid": "container",
		children: [
			/* @__PURE__ */ o("img", {
				onLoad: () => U(),
				alt: x,
				"data-testid": "right-image",
				ref: z,
				src: b,
				style: W.rightImage
			}),
			/* @__PURE__ */ o("img", {
				onLoad: () => U(),
				alt: g,
				"data-testid": "left-image",
				ref: B,
				src: h,
				style: W.leftImage
			}),
			/* @__PURE__ */ s("div", {
				style: W.slider,
				children: [
					/* @__PURE__ */ o("div", { style: W.line }),
					l ? /* @__PURE__ */ o("div", {
						style: W.handleCustom,
						children: l
					}) : /* @__PURE__ */ s("div", {
						style: W.handleDefault,
						children: [/* @__PURE__ */ o("div", { style: W.leftArrow }), /* @__PURE__ */ o("div", { style: W.rightArrow })]
					}),
					/* @__PURE__ */ o("div", { style: W.line })
				]
			}),
			v && /* @__PURE__ */ o("div", {
				style: W.leftLabelContainer,
				children: /* @__PURE__ */ o("div", {
					style: W.leftLabel,
					children: v
				})
			}),
			C && /* @__PURE__ */ o("div", {
				style: W.rightLabelContainer,
				children: /* @__PURE__ */ o("div", {
					style: W.rightLabel,
					children: C
				})
			})
		]
	})] });
};
//#endregion
export { m as default };
